src/app/src/pages/redteam/setup/components/RunOptions.tsx

```diff
@@ -31,6 +44,211 @@ interface RunOptionsProps {
+  const error = isBelowMin(value, 0) ? RUNOPTIONS_TEXT.delayBetweenApiCalls.error : undefined;
+  return (
+    <BaseNumberInput
+      fullWidth
+      label={
+        canSetDelay ? (
+          'Delay between API calls (ms)'
+        ) : (
+          <LabelWithTooltip
+            label="Delay between API calls (ms)"
+            tooltip="To set a delay, you must set the number of concurrent requests to 1."
+          />
+        )
+      }
+      value={value}
+      disabled={!canSetDelay || readOnly}
+      onChange={(v) => {
+        if (readOnly) {
+          return;
+        }
+        setValue(v?.toString() || '');
+      }}
+      onBlur={() => {
+        if (readOnly) {
+          return;
+        }
+        const parsed = value === '' ? 0 : Number(value);
+        const safe = Number.isNaN(parsed) || parsed < 0 ? 0 : parsed;
+        updateRunOption('delay', safe);
+        setValue(String(safe));
+        // Enforce mutual exclusivity
+        updateRunOption('maxConcurrency', 1);
+        setMaxConcurrencyValue('1');
+      }}
+      min={0}
+      slotProps={{
+        input: {
+          readOnly,
+          endAdornment: (
+            <Box sx={{ pl: 1 }}>
+              <Typography variant="caption">ms</Typography>
+            </Box>
+          ),
+        },
+      }}
+      helperText={error || RUNOPTIONS_TEXT.delayBetweenApiCalls.helper}
+      error={Boolean(error)}
+    />
+  );
+};
+
+export interface MaxNumberOfConcurrentRequestsInputProps {
+  value: string;
+  setValue: (value: string) => void;
+  setDelayValue: (value: string) => void;
+  updateRunOption: (key: keyof RedteamRunOptions, value: any) => void;
+  readOnly?: boolean;
+  canSetMaxConcurrency?: boolean;
+}
+
+export const MaxNumberOfConcurrentRequestsInput = ({
+  value,
+  setValue,
+  setDelayValue,
+  updateRunOption,
+  readOnly,
+  canSetMaxConcurrency,
+}: MaxNumberOfConcurrentRequestsInputProps) => {
+  const error = isBelowMin(value, 1) ? RUNOPTIONS_TEXT.maxConcurrentRequests.error : undefined;
+  return (
+    <BaseNumberInput
+      fullWidth
+      label={
+        canSetMaxConcurrency ? (
+          'Max number of concurrent requests'
+        ) : (
+          <LabelWithTooltip
+            label="Max number of concurrent requests"
+            tooltip="To set a max concurrency, you must set the delay to 0."
+          />
+        )
+      }
+      value={value}
+      disabled={!canSetMaxConcurrency || readOnly}
+      onChange={(v) => {
+        if (readOnly) {
+          return;
+        }
+        setValue(v?.toString() || '');
+      }}
+      min={1}
+      onBlur={() => {
+        if (readOnly) {
+          return;
+        }
+        const parsed = value === '' ? 1 : Number(value);
+        const safe = Number.isNaN(parsed) || parsed < 1 ? 1 : parsed;
+        updateRunOption('maxConcurrency', safe);
+        setValue(String(safe));
+        // Enforce mutual exclusivity
+        updateRunOption('delay', 0);
```

Comment:
do the state updates together or try to update both with one function call 

```suggestion
        // Enforce mutual exclusivity
        updateRunOption('delay', 0);
        setValue(String(safe));
```